### PR TITLE
Feature/gh 302 - Add TypeSafety to Product Module

### PR DESCRIPTION
**Pull request by MitchellScottSandre**

### CODE REVIEW

File: projects/core/src/product/occ/product-search.service.spec.ts [hunk 1]
```diff
@@ -5,7 +5,10 @@ import {
 } from '@angular/common/http/testing';
 
 import { OccProductSearchService } from './product-search.service';
-import { OccConfig, SearchConfig } from '@spartacus/core';
+import {
+  OccConfig,
+  SearchConfig
+} from 'projects/backend/occ-client/lib/models';
```
Comment (LTiger14) [inline projects/core/src/product/occ/product-search.service.spec.ts]: Split the two imports and reference the actual place they are from in core. SeachConfig is in product and OccConfig is in occ. ex: import { OccConfig } from '../../occ'; . You could import both from public_api but I still think it's less clean.

File: projects/core/src/product/facade/product-review.service.spec.ts [hunk 2]
```diff
@@ -6,6 +6,7 @@ import { of } from 'rxjs';
 import * as fromStore from '../store';
 
 import { ProductReviewService } from './product-review.service';
+import { Review } from 'projects/backend/occ-client/lib/models';
```
[the remark this answers is not in the document]
Comment (LTiger14) [reply projects/core/src/product/facade/product-review.service.spec.ts]: Another way would be to import it from public_api but I think it's less clean.

File: projects/core/src/product/store/product-state.ts [hunk 3]
```diff
@@ -15,13 +20,13 @@ export interface ProductsState {
...
+  results: ProductList;
+  suggestions: Suggestion[];
+  auxResults: ProductList;
   loading: boolean;
 }
 
 export interface ProductReviewsState {
   productCode: string;
-  list: any[];
+  list: ReviewList;
```
Comment (LTiger14) [inline projects/core/src/product/store/product-state.ts]: I understand what you are thinking but with our current implementation this is a Review[]. To make it a ReviewList you would have to make a few other changes and I think it's easier to make this a review and change the few other types that will also have to change.

File: projects/core/src/product/store/reducers/product-reviews.reducer.ts [hunk 4]
```diff
@@ -3,7 +3,7 @@ import { ProductReviewsState } from '../product-state';
 
 export const initialState: ProductReviewsState = {
   productCode: '',
-  list: []
+  list: { reviews: [] }
```
Comment (LTiger14) [inline projects/core/src/product/store/reducers/product-reviews.reducer.ts]: []

File: projects/core/src/product/store/selectors/product-reviews.selectors.ts [hunk 5]
```diff
@@ -14,7 +14,7 @@ export const getProductReviewsState: MemoizedSelector<
 
 export const getSelectedProductReviewsFactory = (
   productCode
-): MemoizedSelector<any, any> => {
+): MemoizedSelector<any, ReviewList> => {
```
Comment (LTiger14) [inline projects/core/src/product/store/selectors/product-reviews.selectors.ts]: Review[]

File: projects/core/src/product/store/effects/product-reviews.effect.spec.ts [hunk 6]
```diff
@@ -58,7 +58,7 @@ describe('Product reviews effect', () => {
       const action = new fromActions.LoadProductReviews(productCode);
       const completion = new fromActions.LoadProductReviewsSuccess({
         productCode,
-        list: reviewData.reviews
+        list: reviewData
```
Comment (LTiger14) [inline projects/core/src/product/store/effects/product-reviews.effect.spec.ts]: Go back to `reviewData.reviews`

File: projects/core/src/product/store/actions/product-reviews.action.spec.ts [hunk 7]
```diff
@@ -1,6 +1,7 @@
 import * as fromActions from './product-reviews.action';
+import { ErrorModel } from 'projects/backend/occ-client/lib/models';
 
-describe('Product Review Actions', () => {
+fdescribe('Product Review Actions', () => {
```
Comment (LTiger14) [inline projects/core/src/product/store/actions/product-reviews.action.spec.ts]: Please remove f

File: projects/core/src/product/facade/product-review.service.spec.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (MitchellScottSandre) [reply projects/core/src/product/facade/product-review.service.spec.ts]: Thanks for the comments. Will use the relative paths.
But Nick said, in the initial ticket, to "use the models from #301. And 301 refers to models inside of "projects/backend/occ-client/lib/models/index.ts". That's why I used those models.

[the remark this answers is not in the document]
Comment (MitchellScottSandre) [reply projects/core/src/product/facade/product-review.service.spec.ts]: Have fixed all import issues now

File: projects/core/src/product/store/reducers/product-reviews.reducer.spec.ts [hunk 8]
```diff
@@ -1,5 +1,6 @@
 import * as fromReducer from './../reducers/product-reviews.reducer';
```
Comment (LTiger14) [inline projects/core/src/product/store/reducers/product-reviews.reducer.spec.ts]: Can't comment there. On the non test version of this file please add this line 30: 
```
export const getReviewList = (state: ProductReviewsState): Review[] =>
  state.list;
```

This is a fix for an issue with Typescript version 2.9

File: projects/core/src/product/store/reducers/product-reviews.reducer.spec.ts [hunk 9]
```diff
@@ -15,16 +16,18 @@ describe('Product Reviews reducer', () => {
...
-        {
-          id: 1,
-          rating: 3
-        },
-        {
-          id: 2,
-          rating: 5
-        }
-      ];
+      const reviews: ReviewList = {
```
Comment (LTiger14) [inline projects/core/src/product/store/reducers/product-reviews.reducer.spec.ts]: Not a ReviewList but a Review[]

File: projects/core/src/product/store/reducers/product-search.reducer.ts [hunk 10]
```diff
@@ -36,7 +37,7 @@ export function reducer(
     }
 
     case fromProductsSearch.GET_PRODUCT_SUGGESTIONS_SUCCESS: {
-      const suggestions = action.payload;
+      const suggestions: Suggestion[] = action.payload;
 
       return {
         ...state,
```
Comment (LTiger14) [inline projects/core/src/product/store/reducers/product-search.reducer.ts]: Line 55: 
```
export const getSearchResults = (
  state: ProductsSearchState
): ProductSearchPage => state.results;
```
line 58:
```
export const getSearchResults = (
  state: ProductsSearchState
): ProductSearchPage => state.results;
```

Comment (MitchellScottSandre) [reply projects/core/src/product/store/reducers/product-search.reducer.ts]: @LTiger14 the Line 55 changes causes errors. `type Review[] has no properties in common with ProductSearchPage`

File: projects/core/src/product/store/reducers/product-search.reducer.ts [hunk 11]
```diff
@@ -24,7 +25,7 @@ export function reducer(
       const results = action.payload;
 
       if (state.loading) {
-        const res = action.auxiliary ? { auxResults: results } : { results };
+        const res = action.auxiliary ? { auxResults: results } : results;
```
Comment (LTiger14) [inline projects/core/src/product/store/reducers/product-search.reducer.ts]: Put this back const res = action.auxiliary ? { auxResults: results } : { results };

File: projects/core/src/product/facade/product-review.service.spec.ts [hunk 12]
```diff
@@ -40,7 +41,7 @@ describe('ReviewService', () => {
         of(mockReview)
```
Comment (LTiger14) [inline projects/core/src/product/facade/product-review.service.spec.ts]: [mockReview]

File: projects/core/src/product/store/actions/product-reviews.action.spec.ts [hunk 13]
```diff
@@ -26,22 +27,27 @@ describe('Product Review Actions', () => {
...
+              id: '2',
               rating: 5
             }
           ]
         };
-        const action = new fromActions.LoadProductReviewsSuccess(payload);
+
+        const action = new fromActions.LoadProductReviewsSuccess({
+          productCode,
+          list
```
Comment (LTiger14) [inline projects/core/src/product/store/actions/product-reviews.action.spec.ts]: list: list.reviews

File: projects/core/src/product/store/actions/product-reviews.action.spec.ts [hunk 14]
```diff
@@ -26,22 +27,27 @@ describe('Product Review Actions', () => {
...
+        });
         expect({ ...action }).toEqual({
           type: fromActions.LOAD_PRODUCT_REVIEWS_SUCCESS,
-          payload
+          payload: { productCode, list }
```
Comment (LTiger14) [inline projects/core/src/product/store/actions/product-reviews.action.spec.ts]: list: list.reviews

File: projects/core/src/product/store/actions/product-search.action.ts [hunk 15]
```diff
@@ -21,12 +22,12 @@ export class SearchProducts implements Action {
...
 export class SearchProductsFail implements Action {
   readonly type = SEARCH_PRODUCTS_FAIL;
-  constructor(public payload: any, public auxiliary?: boolean) {}
+  constructor(public payload: ErrorModel, public auxiliary?: boolean) {}
 }
 
 export class SearchProductsSuccess implements Action {
   readonly type = SEARCH_PRODUCTS_SUCCESS;
-  constructor(public payload: any, public auxiliary?: boolean) {}
+  constructor(public payload: ProductList, public auxiliary?: boolean) {}
```
Comment (LTiger14) [inline projects/core/src/product/store/actions/product-search.action.ts]: This is a ProductSearchPage

File: projects/core/src/product/store/reducers/product-search.reducer.ts (hunk 10 above)
Comment (MitchellScottSandre) [reply projects/core/src/product/store/reducers/product-search.reducer.ts]: 😉😉😉😉😉😉😉

File: projects/core/src/product/store/reducers/product-search.reducer.ts [hunk 16]
```diff
@@ -1,8 +1,9 @@
 import * as fromProductsSearch from '../actions/product-search.action';
+import { Suggestion, ProductSearchPage } from '../../../occ-models';
 import { ProductsSearchState } from '../product-state';
 
 export const initialState: ProductsSearchState = {
-  results: {},
+  results: [],
```
[the remark this answers is not in the document]
Comment (MitchellScottSandre) [reply projects/core/src/product/store/reducers/product-search.reducer.ts]: thanks

File: projects/core/src/product/facade/product-review.service.spec.ts [hunk 17]
```diff
@@ -37,10 +38,10 @@ describe('ReviewService', () => {
   describe('getByProductCode(productCode)', () => {
     it('should be able to get product reviews if reviews exist', () => {
       spyOnProperty(ngrxStore, 'select').and.returnValue(() => () =>
-        of(mockReview)
+        of([mockReview])
       );
       service.getByProductCode('testId').subscribe(reviews => {
-        expect(reviews).toBe(mockReview);
+        expect(reviews).toBe([mockReview]);
```
Comment (LTiger14) [inline projects/core/src/product/facade/product-review.service.spec.ts]: ToBe won't work for this, put toEqual here.

File: projects/core/src/product/store/selectors/product-reviews.selectors.ts [hunk 18]
```diff
@@ -1,5 +1,5 @@
 import { createSelector, MemoizedSelector } from '@ngrx/store';
```
Comment (LTiger14) [inline projects/core/src/product/store/selectors/product-reviews.selectors.ts]: Please prettify that file (not passing ci)

File: projects/core/src/product/store/reducers/product-reviews.reducer.ts [hunk 19]
```diff
@@ -1,5 +1,6 @@
 import * as fromProductReviews from './../actions/product-reviews.action';
 import { ProductReviewsState } from '../product-state';
+import { Review } from 'projects/core/src/occ-models';
```
Comment (LTiger14) [inline projects/core/src/product/store/reducers/product-reviews.reducer.ts]: Relative path

**Pull request merged**